ffi/tests/firewood/merkle_compatibility_test.go

```diff
@@ -0,0 +1,442 @@
+package tests
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"path/filepath"
+	"slices"
+	"testing"
+
+	firewood "github.com/ava-labs/firewood-go/ffi"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/database/memdb"
+	"github.com/ava-labs/avalanchego/x/merkledb"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	keySize      = 4
+	valSize      = 4
+	maxBatchSize = 5
+	maxNumKeys   = 10_000
+)
+
+const (
+	dbGet byte = iota
+	dbUpdate
+	dbBatch
+	checkDBHash
+	createProposalOnProposal
+	createProposalOnDB
+	proposalGet
+	commitProposal
+	maxStep
+)
+
+var stepMap = map[byte]string{
+	dbGet:                    "dbGet",
+	dbUpdate:                 "dbUpdate",
+	dbBatch:                  "dbBatch",
+	checkDBHash:              "checkDBHash",
+	createProposalOnProposal: "createProposalOnProposal",
+	createProposalOnDB:       "createProposalOnDB",
+	proposalGet:              "proposalGet",
+	commitProposal:           "commitProposal",
+}
+
+func newTestDatabase(t *testing.T) *firewood.Database {
+	t.Helper()
+
+	dbFile := filepath.Join(t.TempDir(), "test.db")
+	db, closeDB, err := newDatabase(dbFile)
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		require.NoError(t, closeDB())
+	})
+	return db
+}
+
+func newDatabase(dbFile string) (*firewood.Database, func() error, error) {
+	conf := firewood.DefaultConfig()
+	conf.MetricsPort = 0
+	conf.Create = true
+
+	f, err := firewood.New(dbFile, conf)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create new database at filepath %q: %w", dbFile, err)
+	}
+	return f, f.Close, nil
+}
+
+type tree struct {
+	require *require.Assertions
+	rand    *rand.Rand
+
+	id       int
+	nextID   int
+	merkleDB merkledb.MerkleDB
```

Comment:
Why is firewood a pointer but not MerkleDB?

Reply:
MerkleDB is an interface